lib/ansible/module_utils/common/parameters.py

```diff
@@ -129,11 +129,23 @@ def list_deprecations(argument_spec, params):
 
     deprecations = []
     for arg_name, arg_opts in argument_spec.items():
-        if arg_opts.get('removed_in_version') is not None and arg_name in params:
-            deprecations.append({
-                'msg': "Param '%s' is deprecated. See the module docs for more information" % arg_name,
-                'version': arg_opts.get('removed_in_version')
-            })
+        if arg_name in params:
+            if arg_opts.get('removed_in_version') is not None:
+                deprecations.append({
+                    'msg': "Param '%s' is deprecated. See the module docs for more information" % (prefix + arg_name),
+                    'version': arg_opts.get('removed_in_version')
+                })
+            # Check sub-argument spec
+            sub_argument_spec = arg_opts.get('options')
+            if sub_argument_spec is not None:
+                sub_arguments = params[arg_name]
+                if isinstance(sub_arguments, Mapping):
+                    sub_arguments = [sub_arguments]
+                if isinstance(sub_arguments, list):
+                    sub_prefix = prefix + arg_name + "' -> '"
```

Comment:
A couple things on this. First, use string formatting rather concatenation since string formatting is faster (`O(n)` vs `O(n**2)`). I don't foresee this being a huge performance problem, but better safe than sorry.
```suggestion
                    sub_prefix = "%s%s ->" % (prefix, arg_name)
```

Second (and this is pretty minor) I'm not sure about using `->` to indicate nested keys. I think it would be better to use standard dictionary syntax, such as `param['subparam']['subsubparam']` if it's not too much trouble.

Reply:
It's only `O(n**2)` when concatenating `O(n)` strings, for concatenating three strings it's still `O(n)`, the difference is only the constant ;) But yes, I'll change it.

About using `->`: I've used it cause validate-modules is also using that. Another alternative would be adding `.` inbetween, which is more concise, but also potentially confusing. I've added commit e0aced4937ab96a52d202702e8abc64010305830 which uses your notation, it now looks like:
```
[DEPRECATION WARNING]: Param 'x' is deprecated. See the module docs for more information. This feature will be removed in version 2.11. Deprecation warnings 
can be disabled by setting deprecation_warnings=False in ansible.cfg.
[DEPRECATION WARNING]: Param 'a["b"]' is deprecated. See the module docs for more information. This feature will be removed in version 2.10. Deprecation 
warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
```
(I've used `"` instead of `'` since the whole thing is already wrapped in `'`s)

Reply:
> It's only O(n**2) when concatenating O(n) strings, for concatenating three strings it's still O(n), the difference is only the constant

Alright, you win the math trophy. 😉 

I think the new format is better since it's syntactically correct. Very nice.

Reply:
Geez, I just realized you've done postdoc work in mathematics 🤦‍♂ . I hope I didn't offend you with my meager attempt at using math. 😊